skrub/_multi_agg_joiner.py

```diff
@@ -0,0 +1,491 @@
+"""
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils.validation import check_is_fitted
+
+from skrub import _join_utils
+from skrub._agg_joiner import AggJoiner
+from skrub._dataframe._namespace import is_pandas, is_polars
+from skrub._utils import _is_array_like, atleast_2d_or_none
+
+
+class MultiAggJoiner(TransformerMixin, BaseEstimator):
+    """Extension of the AggJoiner to multiple auxiliary tables.
+
+    Apply numerical and categorical aggregation operations on the `cols`
+    to aggregate, selected by dtypes. See the list of supported operations
+    at the parameter `operations`.
+
+    If `cols` is not provided, `cols` is set to a list of lists.
+    For each table in `aux_tables`, the correponding list will be all columns
+    of that table, except the `aux_keys` associated with that table.
+
+    As opposed to the :class:`AggJoiner`, here `aux_tables` is a list of tables,
+    each of which will be joined on the main table. Therefore `aux_keys` is now
+    a list of keys, of the same length as `aux_tables`, and each entry in `aux_keys`
+    is used to join the corresponding aux table. In the same way, each entry in `cols`
+    is a list of columns to aggregate in the corresponding aux table.
+
+    Therefore if we have a single table, we could either use
+
+    - the AggJoiner: ``AggJoiner(aux_table, key="ID")``
+    - or the MultiAggJoiner: ``MultiAggJoiner([aux_table], keys=["ID"])``
+
+    Note that for `keys`, `main_keys`, `aux_keys`, `cols` and `operations`,
+    an input of the form ``[["a"], ["b"], ["c", "d"]]`` is valid
+    while ``["a", "b", ["c", "d"]]`` is not.
+
+    Using a column from the first aux table to join the second aux table is not
+    (yet) supported.
+
+    Accepts :obj:`pandas.DataFrame` and :class:`polars.DataFrame` inputs.
+
+    Parameters
+    ----------
+    aux_tables : iterable of DataFrameLike or "X"
+        Auxiliary dataframes to aggregate then join on the base table.
+        The placeholder string "X" can be provided to perform
+        self-aggregation on the input data. To provide a single auxiliary table,
+        ``aux_tables = [table]`` is supported, but not ``aux_tables = table``.
+        It's possible to provide both the placeholder "X" and auxiliary tables,
+        as in ``aux_tables = [table, "X"]``. If that's the case, the second table will
+        be replaced by the input data.
+
+    keys : iterable of str, or iterable of iterable of str, default=None
+        The column names to use for both `main_keys` and `aux_key` when they
+        are the same. Provide either `key` or both `main_keys` and `aux_keys`.
+        If there are multiple auxiliary tables, `keys` will be used to join all
+        of them.
+
+        If not `None`, there must be an iterable of `keys` for each table
+        in `aux_tables`.
+
+    main_keys : iterable of str, or iterable of iterable of str, default=None
```

Comment:
can you say that they must be in the main table, ie we cannot use columns from the first joined table to join the second